### DISCUSSION TITLE
C++ 静态反射与序列化 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by giscus** [SF-Zhou/sf-zhou.github.io#198] [Announcements]
# C++ 静态反射与序列化 | SF-Zhou's Blog

https://sf-zhou.github.io/programming/cpp_static_reflection.html

**Comment by SF-Zhou** [comment]
文中用到的 CRTP 的核心原因是想拿到当前类型，然而已经有黑科技可以通过 ADL 拿到当前的类型了：[MitalAshok/self_macro](https://github.com/MitalAshok/self_macro)，感叹 C++ 开发者们的奇思妙想。一个简短的示例：https://stackoverflow.com/questions/21143835/can-i-implement-an-autonomous-self-member-type-in-c/70701479#70701479

**Reply by SF-Zhou** [reply]
文中代码更新为使用 ADL 获取当前类型，不再需要 CRTP 了。

**Reply by SF-Zhou** [reply]
一件有意思的事，类型加上模板后 clang 无法编译超过 3 个字段的情况，换成 gcc 则没问题，https://godbolt.org/z/6bnqnj71q

**Reply by SF-Zhou** [reply]
该问题可以通过将 REFL_NOW 抽出来先定义的方式规避，https://godbolt.org/z/5eo8Ts3rY

**Reply by SF-Zhou** [reply]
ADL 也并非是必需品，用 lambda 直接返回字段亦可，https://godbolt.org/z/K8Yz6eT5c

**Comment by SF-Zhou** [comment]
C++ 的发展日新月异，C++23 已经可以实现类似 Rust 的 borrow checker 了（[链接](https://www.reddit.com/r/cpp/comments/10zxutf/fullfledged_affine_type_destructive_move_in_c23/)），有空的时候我研究下实现。

**Reply by SF-Zhou** [reply]
核心原理（[链接](https://godbolt.org/z/W8Kjhczer)）：
```c++
#include <concepts>
#include <type_traits>
#include <utility>

template <auto, auto>
struct ADLType {
 friend consteval auto ADLFunc(ADLType, auto...);
};

template <auto InstanceIdentifier, auto Rank>
struct InjectDefinitionForADLFunc {
 inline static constexpr auto rank = Rank;
 friend consteval auto ADLFunc(ADLType<InstanceIdentifier, Rank>, auto...) {}
};

template <auto InstanceIdentifier, auto Rank = 0, auto = [] {}>
consteval auto ExtractThenUpdateCurrentState() -> decltype(Rank) {
 if constexpr (requires { ADLFunc(ADLType<InstanceIdentifier, Rank>{}); })
   return ExtractThenUpdateCurrentState<InstanceIdentifier, Rank + 1>();
 else
   return InjectDefinitionForADLFunc<InstanceIdentifier, Rank>::rank;
}

constexpr auto Identifier = [] {};
static_assert(ExtractThenUpdateCurrentState<Identifier>() == 0);
static_assert(ExtractThenUpdateCurrentState<Identifier>() == 1);

int main() { return 0; }
```

**Comment by SF-Zhou** [comment]
[boost::pfr](https://github.com/boostorg/pfr/releases/tag/2.2.0) 实现了更优雅的 [get_name](https://github.com/boostorg/pfr/pull/129/files)，底层原理依然是解析 `__PRETTY_FUNCTION__`。本文提到的方法意义不大了。
